src/spi.rs

```diff
@@ -0,0 +1,495 @@
+             .ssi().set_bit()
+             .ssm().set_bit()
+             .mstr().set_bit()
+        });
+    }
+
+    ///Configures CR2 register
+    fn configure_cr2(&self) {
+        self.cr2().write(|w| unsafe {
+            //Data size 8 bit
+            w.ds().bits(0b111)
+             .ssoe().set_bit()
+             //RXNE event is generated if the FIFO level is greater than or equal to 1/4 (8-bit)
+             .frxth().set_bit()
+        });
+    }
+
+    ///Enables SPI
+    fn enable(apb: &mut Self::APB);
+}
+
+impl InnerSpi for SPI1 {
+    const IDX: u8 = 1;
+    type APB = APB2;
+
+    #[inline]
+    fn get_clock_freq(clocks: &Clocks) -> Hertz {
+        clocks.pclk2()
+    }
+
+    fn cr1(&self) -> &::stm32l4x6::spi1::CR1 {
+        &self.cr1
+    }
+
+    fn cr2(&self) -> &::stm32l4x6::spi1::CR2 {
+        &self.cr2
+    }
+
+    fn sr(&self) -> &::stm32l4x6::spi1::SR {
+        &self.sr
+    }
+
+    fn dr(&self) -> &::stm32l4x6::spi1::DR {
+        &self.dr
+    }
+
+    fn enable(apb: &mut Self::APB) {
+        // enable and/or reset SPI
+        apb.enr().modify(|_, w| w.spi1en().set_bit());
+        apb.rstr().modify(|_, w| w.spi1rst().set_bit());
+        apb.rstr().modify(|_, w| w.spi1rst().clear_bit());
+    }
+}
+
+impl InnerSpi for SPI2 {
+    const IDX: u8 = 2;
+    type APB = APB1;
+
+    #[inline]
+    fn get_clock_freq(clocks: &Clocks) -> Hertz {
+        clocks.pclk2()
+    }
+
+    fn cr1(&self) -> &::stm32l4x6::spi1::CR1 {
+        &self.cr1
+    }
+
+    fn cr2(&self) -> &::stm32l4x6::spi1::CR2 {
+        &self.cr2
+    }
+
+    fn sr(&self) -> &::stm32l4x6::spi1::SR {
+        &self.sr
+    }
+
+    fn dr(&self) -> &::stm32l4x6::spi1::DR {
+        &self.dr
+    }
+
+    fn enable(apb: &mut Self::APB) {
+        // enable and/or reset SPI
+        apb.enr1().modify(|_, w| w.spi2en().set_bit());
+        apb.rstr1().modify(|_, w| w.spi2rst().set_bit());
+        apb.rstr1().modify(|_, w| w.spi2rst().clear_bit());
+    }
+}
+
+impl InnerSpi for SPI3 {
+    const IDX: u8 = 3;
+    type APB = APB1;
+
+    #[inline]
+    fn get_clock_freq(clocks: &Clocks) -> Hertz {
+        clocks.pclk2()
+    }
+
+    fn cr1(&self) -> &::stm32l4x6::spi1::CR1 {
+        &self.cr1
+    }
+
+    fn cr2(&self) -> &::stm32l4x6::spi1::CR2 {
+        &self.cr2
+    }
+
+    fn sr(&self) -> &::stm32l4x6::spi1::SR {
+        &self.sr
+    }
+
+    fn dr(&self) -> &::stm32l4x6::spi1::DR {
+        &self.dr
+    }
+
+    fn enable(apb: &mut Self::APB) {
+        // enable and/or reset SPI
+        apb.enr1().modify(|_, w| w.sp3en().set_bit());
+        apb.rstr1().modify(|_, w| w.spi3rst().set_bit());
+        apb.rstr1().modify(|_, w| w.spi3rst().clear_bit());
+    }
+}
+
+
+/// SPI errors.
+pub enum Error {
+    /// Overrun occurred
+    Overrun,
+    /// Mode fault
+    ModeFault,
+    /// CRC cheksum error.
+    Crc,
+}
+
+/// SPI
+pub struct Spi<SPI, SCK, MISO, MOSI> {
+    spi: SPI,
+    pins: (SCK, MISO, MOSI),
```

Comment:
Because you don't actually operate on the pins themselves (you just consume them at `Spi` creation time), the type signature can be simplified to just
```
pub struct Spi<SPI, PINS> {
    spi: SPI,
    pins: PINS,
}
```
and your impl of `FullDuplex` looks like
```
impl<PINS> FullDuplex<u8> for Spi<SPIx, PINS> {
//...
}
```
(see [example](https://github.com/japaric/stm32f30x-hal/blob/master/src/spi.rs#L156))

Reply:
But I would not be able to do trait bounds on impl block of Spi